xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1354 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package gracefulswitch
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+const (
+	defaultTestTimeout      = 5 * time.Second
+	defaultTestShortTimeout = 10 * time.Millisecond
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func setup(t *testing.T) (*testutils.TestClientConn, *Balancer) {
+	tcc := testutils.NewTestClientConn(t)
+	return tcc, NewGracefulSwitchBalancer(tcc, balancer.BuildOptions{})
+}
+
+// TestSuccessfulFirstUpdate tests a basic scenario for the Graceful Switch Load
+// Balancer, where it is setup with a balancer which should populate the Current
+// Load Balancer. Any ClientConn updates should then be forwarded to this
+// Current Load Balancer.
+func (s) TestSuccessfulFirstUpdate(t *testing.T) {
+	_, gsb := setup(t)
+	builder := balancer.Get(balancerName1)
+	if builder == nil {
+		t.Fatalf("balancer.Get(%v) returned nil", balancerName1)
+	}
+	if err := gsb.SwitchTo(builder); err != nil {
+		t.Fatalf("Balancer.SwitchTo failed with error: %v", err)
+	}
+	if gsb.balancerCurrent == nil {
+		t.Fatal("balancerCurrent was not built out when a correct SwitchTo() call should've triggered the balancer to build")
+	}
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to the Current Load Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancer1Config{},
+	}
+
+	// Updating ClientConnState should forward the update exactly as is to the
+	// current balancer.
+	if err := gsb.UpdateClientConnState(ccs); err != nil {
+		t.Fatalf("Balancer.UpdateClientConnState(%v) failed with error: %v", ccs, err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
```

Comment:
Why the short timeout here?  I think this will be flaky.

The reason for the short timeout is if you are expecting a failure to occur due to the timeout, you do not want to wait 5 seconds.  Not because you expect something to be fast.  The VMs we run our tests on hiccup sometimes.  It looks like your other usages are correct, so maybe this was just an oversight.